### ISSUE TITLE
Cell type NoData logic for unsigned byte / short not working properly

### ISSUE THREAD

**Issue by lossyrob** [locationtech/geotrellis#2166]
```scala
scala> import geotrellis.raster.{CellType, UByteUserDefinedNoDataCellType}
import geotrellis.raster.{CellType, UByteUserDefinedNoDataCellType}

scala> val ct = UByteUserDefinedNoDataCellType(255.toByte)
ct: geotrellis.raster.UByteUserDefinedNoDataCellType = uint8ud-1

scala> ct.toString
res1: String = uint8ud-1

scala> CellType.fromString("uint8ud255")
<console>:22: warning: method fromString in object CellType is deprecated: Use `fromName` instead
       CellType.fromString("uint8ud255")
                ^
java.lang.NumberFormatException: Value out of range. Value:"255" Radix:10
  at java.lang.Byte.parseByte(Byte.java:151)
  at java.lang.Byte.parseByte(Byte.java:175)
  at scala.collection.immutable.StringLike$class.toByte(StringLike.scala:264)
  at scala.collection.immutable.StringOps.toByte(StringOps.scala:29)
  at geotrellis.raster.CellType$.fromName(CellType.scala:428)
  at geotrellis.raster.CellType$.fromString(CellType.scala:395)
  ... 42 elided

scala> 
````

Reported by @metasim

**Comment by pomadchin** [comment]
Unsigned short has the same issue: 

```
scala> import geotrellis.raster._
import geotrellis.raster._

scala> val ct = UShortUserDefinedNoDataCellType(65535.toShort)
ct: geotrellis.raster.UShortUserDefinedNoDataCellType = uint16ud-1

scala> ct.toString
res0: String = uint16ud-1

scala> CellType.fromString("uint16ud65535")
java.lang.NumberFormatException: Value out of range. Value:"65535" Radix:10
  at java.lang.Short.parseShort(Short.java:120)
  at java.lang.Short.parseShort(Short.java:144)
  at scala.collection.immutable.StringLike$class.toShort(StringLike.scala:268)
  at scala.collection.immutable.StringOps.toShort(StringOps.scala:29)
  at geotrellis.raster.CellType$.fromString(CellType.scala:457)
  ... 42 elided
```

**Comment by metasim** [comment]
To start off, I think we need a more clear/explicit definition on what the valid value and no-data ranges should be for the CellTypes. Here's a straw man:

```scala
  sealed trait CellDef {
    type Encoding
    val base: String
    val ndRange: Range
  }

  trait UByteDef extends CellDef {
    type Encoding = Short
    val base = "uint8ud"
    val ndRange = 0 to Byte.MaxValue * 2
  }

  trait ByteDef extends CellDef {
    type Encoding = Byte
    val base = "int8ud"
    val ndRange = Byte.MinValue to Byte.MaxValue
  }

  trait UShortDef extends CellDef {
    type Encoding = Char
    val base = "uint16ud"
    val ndRange = 0 to Short.MaxValue * 2
  }

  trait ShortDef extends CellDef {
    type Encoding = Short
    val base = "int16ud"
    val ndRange = Short.MinValue to Short.MaxValue
  }

  trait IntDef extends CellDef {
    type Encoding = Int
    val base = "int32ud"
    val ndRange = Int.MinValue to Int.MaxValue
  }

  trait FloatDef extends CellDef {
    type Encoding = Float
    val base = "float32ud"
    val ndRange = Float.MinValue to Float.MaxValue
  }

  trait DoubleDef extends CellDef {
    type Encoding = Double
    val base = "float64ud"
    val ndRange = Double.MinValue to Double.MaxValue
  }
```

Are the assumptions herein correct? How do these map to the [GeoTIFF spec](https://www.itu.int/itudoc/itu-t/com16/tiff-fx/docs/tiff6.pdf)? Or [C types](https://en.wikipedia.org/wiki/C_data_types#Basic_types)?

**Comment by lossyrob** [comment]
Those assumptions are correct. Is there a reason to make them explicit in code, vs documenting them?

We've always operated under the assumption that the user will handle overflows themselves. If you're working with a Byte raster, and you set the Int value 500, you have to deal with the overflow. In the case of our UByte rasters, if you set 255, you'll get 255 back out. For our NoData handling, though, we didn't support the logic that would allow this, and this is where I see the bug.

**Comment by metasim** [comment]
I was mostly encoding them in "code" form so there was not ambiguity in our dialog over it. That said, I do think there is value in having them in code to facilitate generation of tests that cover the whole bounds/edge cases. Right now I've been revising the above in the `CellTypeSpec` and creating tests that expand on the existing ones to ensure the edge cases are covered from hereon out.

**Comment by metasim** [comment]
Current encoding parser also fails to handle negative ND values:

```scala
scala> import geotrellis.raster._
import geotrellis.raster._

scala> val ct = IntUserDefinedNoDataCellType(-1)
ct: geotrellis.raster.IntUserDefinedNoDataCellType = int32ud-1

scala> val ct2 = CellType.fromName(ct.name)
ct2: geotrellis.raster.CellType = int32ud1
```

**Comment by metasim** [comment]
Note, @mteldridge found another bug associated with writing out TIFF files with no-data values for unsigned values that are outside the signed range. Addressed with 658f956.
